File: Java/src/test/java/org/chess/domain/rules/TargetPositionMustNotBeOccupiedByAllyTest.java
package org.chess.domain.rules;

import org.chess.domain.*;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

class TargetPositionMustNotBeOccupiedByAllyTest {

    @Test
    void shouldRejectMoveToPositionOccupiedByAlly() {
        final var positionFrom = new Position(3, 3);
        final var positionTo = new Position(4, 4);
        final var whiteBishopToMove = new ChessPiece(ChessPieceName.BISHOP, PlayerColor.WHITE);
        final var whitePawn = new ChessPiece(ChessPieceName.PAWN, PlayerColor.WHITE);

        final var gameState = new GameState(Map.of(
                positionFrom, whiteBishopToMove,
                positionTo, whitePawn));
        final var move = new Move(positionFrom, positionTo);

        final var rule = new TargetPositionMustNotBeOccupiedByAlly();
        final var ruleViolation = rule.execute(move, gameState);

        assertThat(ruleViolation).isPresent();
        assertThat(ruleViolation.get()).isEqualTo(RuleViolation.TARGET_POSITION_OCCUPIED_BY_ALLY);
    }

    @Test
    void shouldRejectMoveToEmptyPosition() {
        final var positionFrom = new Position(3, 3);
        final var positionTo = new Position(4, 4);
        final var whiteBishopToMove = new ChessPiece(ChessPieceName.BISHOP, PlayerColor.WHITE);

        final var gameState = new GameState(Map.of(positionFrom, whiteBishopToMove));
        final var move = new Move(positionFrom, positionTo);

        final var rule = new TargetPositionMustNotBeOccupiedByAlly();
        final var ruleViolation = rule.execute(move, gameState);

        assertThat(ruleViolation).isNotPresent();
    }
}

File: Kotlin/src/test/kotlin/org/chess/domain/rules/TargetPositionMustNotBeOccupiedByAllyTest.kt
package org.chess.domain.rules

import org.assertj.core.api.Assertions.*
import org.chess.domain.*
import org.junit.jupiter.api.Test

internal class TargetPositionMustNotBeOccupiedByAllyTest {

    @Test
    fun shouldRejectMoveToPositionOccupiedByAlly() {
        val positionFrom = Position(3, 3)
        val positionTo = Position(4, 4)
        val whiteBishopToMove = ChessPiece(ChessPieceName.BISHOP, PlayerColor.WHITE)
        val whitePawn = ChessPiece(ChessPieceName.PAWN, PlayerColor.WHITE)
        val gameState = GameState(
            mapOf(
                Pair(positionFrom, whiteBishopToMove),
                Pair(positionTo, whitePawn)
            )
        )

        val move = Move(positionFrom, positionTo)
        val rule = TargetPositionMustNotBeOccupiedByAlly()
        val ruleViolation = rule.execute(move, gameState)

        assertThat(ruleViolation).isEqualTo(RuleViolation.TARGET_POSITION_OCCUPIED_BY_ALLY)
    }

    @Test
    fun shouldNotRejectMoveToEmptyPosition() {
        val positionFrom = Position(3, 3)
        val positionTo = Position(4, 4)
        val whiteBishopToMove = ChessPiece(ChessPieceName.BISHOP, PlayerColor.WHITE)
        val gameState = GameState(
            mapOf(
                Pair(positionFrom, whiteBishopToMove)
            )
        )

        val move = Move(positionFrom, positionTo)
        val rule = TargetPositionMustNotBeOccupiedByAlly()
        val ruleViolation = rule.execute(move, gameState)

        assertThat(ruleViolation).isNull()
    }
}
